internal/searchindex/searchindex.go

```diff
@@ -0,0 +1,440 @@
+package searchindex
+
+import (
+	"bytes"
+	"encoding/gob"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/cmp"
+
+	"go.mongodb.org/atlas-sdk/v20231115008/admin"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/compat"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/pointer"
+	akov2 "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1/common"
+)
+
+type SearchIndex struct {
+	akov2.SearchIndex
+	akov2.AtlasSearchIndexConfigSpec
+	ID     *string
+	Status *string
+}
+
+func (s *SearchIndex) Key() string {
+	var buf bytes.Buffer
+	encoder := gob.NewEncoder(&buf)
+	err := encoder.Encode(*s)
+	fmt.Println("ERR:", err)
+	return buf.String()
+}
+
+func (s *SearchIndex) GetID() string {
+	return pointer.GetOrDefault(s.ID, "")
+}
+
+func (s *SearchIndex) GetStatus() string {
+	return pointer.GetOrDefault(s.Status, "")
+}
+
+func NewSearchIndexFromAKO(index *akov2.SearchIndex, config *akov2.AtlasSearchIndexConfigSpec) *SearchIndex {
+	if index == nil || config == nil {
+		return nil
+	}
+	return &SearchIndex{
+		SearchIndex:                *index.DeepCopy(),
+		AtlasSearchIndexConfigSpec: *config.DeepCopy(),
+	}
+}
+
+func NewSearchIndexFromAtlas(index admin.ClusterSearchIndex) (*SearchIndex, error) {
+	convertVectorFields := func(in *[]map[string]interface{}) (*apiextensionsv1.JSON, error) {
+		if in == nil {
+			return nil, nil
+		}
+		result := &apiextensionsv1.JSON{}
+		err := compat.JSONCopy(result, *in)
+		return result, err
```

Comment:
```suggestion
		if err := compat.JSONCopy(result, *in); err != nil {
			return nil, err
		}
		return result, nil
```

Reply:
Why to have two returns? If there is an error, this means the result is invalid.

Reply:
See https://go.dev/doc/tutorial/handle-errors

Reply:
and https://go.dev/doc/effective_go#multiple-returns